File: FSUIKit/Classes/FSDatePickerView.m
//
//  FSDatePickerView.m
//  ShareEconomy
//
//  Created by FudonFuchina on 16/8/7.
//  Copyright © 2016年 FudonFuchina. All rights reserved.
//

#import "FSDatePickerView.h"

@interface FSDatePickerView ()

@property (nonatomic,strong) UIView         *mainView;
@property (nonatomic,strong) UIDatePicker   *datePicker;

@end

@implementation FSDatePickerView{
    NSDate  *_date;
}

- (instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        [self dpDesignViews];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame date:(NSDate *)date{
    self = [super initWithFrame:frame];
    if (self) {
        _date = date;
        [self dpDesignViews];
    }
    return self;
}

- (void)dpDesignViews{
    UIView *backView = [[UIView alloc] initWithFrame:self.bounds];
    backView.backgroundColor = [UIColor blackColor];
    backView.alpha = .28;
    [self addSubview:backView];
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tap)];
    [backView addGestureRecognizer:tap];

    CGSize size = self.bounds.size;
    _mainView = [[UIView alloc] initWithFrame:CGRectMake(0, size.height, size.width, 240)];
    _mainView.backgroundColor = [UIColor whiteColor];
    [self addSubview:_mainView];
    
    _datePicker = [[UIDatePicker alloc] initWithFrame:CGRectMake(0, 40, self.bounds.size.width, 200)];
    _datePicker.datePickerMode = UIDatePickerModeDate;
    [_datePicker setLocale:[NSLocale localeWithLocaleIdentifier:@"zh-CN"]];
    _datePicker.minimumDate = [[NSDate date] dateByAddingTimeInterval:- 100 * 365 * 24 * 3600];
    _datePicker.maximumDate = [[NSDate date] dateByAddingTimeInterval:100 * 365 * 24 * 3600];
//    [_datePicker addTarget:self action:@selector(datePickerValueChanged:) forControlEvents:UIControlEventValueChanged];
    [_mainView addSubview:_datePicker];
    
    if (![_date isKindOfClass:NSDate.class]) {
        _date = [NSDate date];
    }
    _datePicker.date = _date;
    
    UIView *selectView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, _mainView.bounds.size.width, 40)];
    selectView.backgroundColor = [UIColor colorWithRed:18/255.0 green:152/255.0 blue:233/255.0 alpha:1.0f];
    [_mainView addSubview:selectView];
    
    for (int x = 0; x < 2; x ++) {
        UIButton *button = [UIButton buttonWithType:UIButtonTypeSystem];
        button.frame = CGRectMake(x * (selectView.bounds.size.width - 60), 0, 60, 40);
        [button setTitle:x?NSLocalizedString(@"Confirm", nil):NSLocalizedString(@"Cancel", nil) forState:UIControlStateNormal];
        [button setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        button.titleLabel.font = [UIFont systemFontOfSize:15];
        button.tag = x;
        [button addTarget:self action:@selector(buttonAction:) forControlEvents:UIControlEventTouchUpInside];
        [selectView addSubview:button];
    }
    
    NSDateComponents *comp = [FSDatePickerView componentForDate:[NSDate date]];
    NSString *title = [[NSString alloc] initWithFormat:@"今天是%@月%@日",@(comp.month),@(comp.day)];
    UILabel *timeLabel = [[UILabel alloc] initWithFrame:CGRectMake(60, 0, size.width - 120, selectView.bounds.size.height)];
    timeLabel.textColor = [UIColor whiteColor];
    timeLabel.font = [UIFont systemFontOfSize:14];
    timeLabel.textAlignment = NSTextAlignmentCenter;
    timeLabel.text = title;
    [selectView addSubview:timeLabel];
    
    [self showAction:YES];
}

- (void)tap{
    if (self.cancel) {
        self.cancel();
    }
    [self showAction:NO];
}

- (void)buttonAction:(UIButton *)button{
    if (button.tag) {
        if (_block) {
//            NSCalendar *calendar = [NSCalendar currentCalendar];
//            NSDateComponents *components = [calendar components:NSUIntegerMax fromDate:_datePicker.date];
//            components.hour = 0;components.minute = 0;components.second = 0;
//            NSDate *date = [calendar dateFromComponents:components];
            _block(_datePicker.date);
        }
    }
    [self showAction:NO];
}

- (void)showAction:(BOOL)show{
    if (show) {
        [UIView animateWithDuration:.3 animations:^{
            self.mainView.frame = CGRectMake(self.mainView.frame.origin.x, self.frame.size.height - 240, self.mainView.bounds.size.width, self.mainView.bounds.size.height);
        }];
    }else{
        [UIView animateWithDuration:.3 animations:^{
            self.mainView.frame = CGRectMake(self.mainView.frame.origin.x, self.frame.size.height, self.mainView.bounds.size.width, self.mainView.bounds.size.height);
        } completion:^(BOOL finished) {
            [self removeFromSuperview];
        }];
    }
}

+ (NSDateComponents *)componentForDate:(NSDate *)date{
    NSCalendar *calendar = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierGregorian];
    NSDateComponents *components = [calendar components:NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay | NSCalendarUnitHour | NSCalendarUnitMinute | NSCalendarUnitSecond | NSCalendarUnitWeekOfMonth | NSCalendarUnitWeekday | NSCalendarUnitWeekOfMonth | NSCalendarUnitWeekOfYear fromDate:date];
    return components;
}

//- (void)datePickerValueChanged:(UIDatePicker *)datePicker
//{
//    if (_block) {
//        _block(datePicker.date);
//    }
//}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end

File: FSUIKit/Classes/Swift/FSDatePickerView.swift
// FSDatePickerView.swift
// Translated from FSDatePickerView.h/m

import UIKit

public typealias FSDatePickerBlock = (Date) -> Void

public class FSDatePickerView: UIView {
    
    private var mainView: UIView!
    private var datePicker: UIDatePicker!
    private var _date: Date?
    
    public var block: FSDatePickerBlock?
    public var cancel: (() -> Void)?
    
    public convenience init(frame: CGRect, date: Date?) {
        self.init(frame: frame)
        _date = date
        dpDesignViews()
    }
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        dpDesignViews()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        dpDesignViews()
    }
    
    private func dpDesignViews() {
        let backView = UIView(frame: bounds)
        backView.backgroundColor = .black
        backView.alpha = 0.28
        addSubview(backView)
        
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(tapAction))
        backView.addGestureRecognizer(tapGesture)
        
        let size = bounds.size
        mainView = UIView(frame: CGRect(x: 0, y: size.height, width: size.width, height: 240 + 34))
        mainView.backgroundColor = .white
        addSubview(mainView)
        
        datePicker = UIDatePicker()
        datePicker.datePickerMode = .date
        if #available(iOS 13.4, *) {
            datePicker.preferredDatePickerStyle = .wheels
        }
        datePicker.frame = CGRect(x: 0, y: 40, width: bounds.width, height: 200)
        datePicker.locale = Locale(identifier: "zh-CN")
        datePicker.minimumDate = Date().addingTimeInterval(-100 * 365 * 24 * 3600)
        datePicker.maximumDate = Date().addingTimeInterval(100 * 365 * 24 * 3600)
        mainView.addSubview(datePicker)
        
        if let date = _date {
            datePicker.date = date
        } else {
            datePicker.date = Date()
        }
        
        let selectView = UIView(frame: CGRect(x: 0, y: 0, width: mainView.bounds.width, height: 40))
        selectView.backgroundColor = UIColor(red: 18/255.0, green: 152/255.0, blue: 233/255.0, alpha: 1.0)
        mainView.addSubview(selectView)
        
        for x in 0..<2 {
            let button = UIButton(type: .system)
            button.frame = CGRect(x: CGFloat(x) * (selectView.bounds.width - 60), y: 0, width: 60, height: 40)
            button.setTitle(x == 1 ? "确认" : "取消", for: .normal)
            button.setTitleColor(.white, for: .normal)
            button.titleLabel?.font = UIFont.systemFont(ofSize: 15)
            button.tag = x
            button.addTarget(self, action: #selector(buttonAction(_:)), for: .touchUpInside)
            selectView.addSubview(button)
        }
        
        let comp = Self.component(for: Date())
        let title = "今天是\(comp.month ?? 0)月\(comp.day ?? 0)日"
        let timeLabel = UILabel(frame: CGRect(x: 60, y: 0, width: size.width - 120, height: selectView.bounds.height))
        timeLabel.textColor = .white
        timeLabel.font = UIFont.systemFont(ofSize: 14)
        timeLabel.textAlignment = .center
        timeLabel.text = title
        selectView.addSubview(timeLabel)
        
        showAction(true)
    }
    
    @objc private func tapAction() {
        cancel?()
        showAction(false)
    }
    
    @objc private func buttonAction(_ button: UIButton) {
        if button.tag == 1 {
            block?(datePicker.date)
            showAction(false)
        } else {
            tapAction()
        }
    }
    
    private func showAction(_ show: Bool) {
        if show {
            UIView.animate(withDuration: 0.3) {
                self.mainView.frame = CGRect(x: self.mainView.frame.origin.x, y: self.frame.height - 240 - 34, width: self.mainView.bounds.width, height: self.mainView.bounds.height)
            }
        } else {
            UIView.animate(withDuration: 0.3, animations: {
                self.mainView.frame = CGRect(x: self.mainView.frame.origin.x, y: self.frame.height, width: self.mainView.bounds.width, height: self.mainView.bounds.height)
            }, completion: { _ in
                self.removeFromSuperview()
            })
        }
    }
    
    public static func component(for date: Date) -> DateComponents {
        let calendar = Calendar(identifier: .gregorian)
        return calendar.dateComponents([.year, .month, .day, .hour, .minute, .second, .weekOfMonth, .weekday, .weekOfYear], from: date)
    }
}
